### PR TITLE
Parse Swift decls in one shot

### PR DESCRIPTION
**Pull request by hamishknight**
This is an intermediate step to requestifying `SourceFile::getTopLevelDecls`. Instead of interleaving type-checking and parsing for SIL files, first parse the file for Swift decls by skipping over any intermixed SIL decls. Then we can perform type checking, and finally SIL parsing where we now skip over Swift decls.

I decided to go for this approach over requiring Swift decls to precede SIL decls as we have ~500 tests with intermixed Swift and SIL, and they are a bit more readable when the Swift code is next the SIL being tested.

### CODE REVIEW

**Comment by hamishknight** [comment]
@swift-ci please test

**Comment by hamishknight** [comment]
@swift-ci please test source compatibility

**Comment by hamishknight** [comment]
@swift-ci please smoke test compiler performance

**Comment by hamishknight** [comment]
https://github.com/apple/llvm-project/pull/712

@swift-ci please test

File: lib/Parse/ParseDecl.cpp [hunk 1]
```diff
@@ -3347,7 +3371,35 @@ bool Parser::isStartOfDecl() {
     if (Tok.isAny(tok::r_brace, tok::eof, tok::pound_endif))
       return true;
 
-    return isStartOfDecl();
+    return isStartOfSwiftDecl();
+  }
+
+  // If we have a decl modifying keyword, check if the next token is a valid
+  // decl start. This is necessary to correctly handle Swift keywords that are
+  // shared by SIL, e.g 'private'.
```
Comment (slavapestov) [inline lib/Parse/ParseDecl.cpp]: Can you give an example?

**Review by slavapestov — APPROVED** [review]
This is a really clever solution to the problem of intermingled AST and SIL declarations. Nice job!

File: lib/Frontend/Frontend.cpp [hunk 2]
```diff
@@ -953,6 +947,7 @@ bool CompilerInstance::parsePartialModulesAndLibraryFiles(
 
 void CompilerInstance::parseAndTypeCheckMainFileUpTo(
```
Comment (slavapestov) [inline lib/Frontend/Frontend.cpp]: At this point, there probably isn't anything special about the MainBufferID remaining. You might want to try removing this function altogether, and handling the main buffer uniformly with all the others.

File: lib/Parse/ParseDecl.cpp (hunk 1 above)
Comment (hamishknight) [reply lib/Parse/ParseDecl.cpp]: @slavapestov This handles cases like `sil private @foo : @convention(thin) () -> ()`. If we unconditionally treat `private` as the start of a decl, we'll try to parse this as a Swift decl. Should I add this example to the comment?

File: lib/Frontend/Frontend.cpp (hunk 2 above)
Comment (hamishknight) [reply lib/Frontend/Frontend.cpp]: @slavapestov Good idea! I'll have a go at that as a follow-up.

Thanks for the review!

**Review by rintaro — APPROVED** [review]

File: lib/Parse/ParseDecl.cpp [hunk 3]
```diff
@@ -3383,7 +3435,29 @@ bool Parser::isStartOfDecl() {
...
+  case tok::kw_sil:
+  case tok::kw_sil_stage:
+  case tok::kw_sil_property:
+  case tok::kw_sil_vtable:
+  case tok::kw_sil_global:
+  case tok::kw_sil_witness_table:
+  case tok::kw_sil_default_witness_table:
+  case tok::kw_sil_coverage_map:
+  case tok::kw_sil_scope:
+    return true;
```
Comment (rintaro) [inline lib/Parse/ParseDecl.cpp]: I believe that SIL keywords are escaped if developers use them in their swift files (e.g. `struct sil {}` becomes ``struct `sil` {}`` in SIL), but for *extra* safety, how about checking `isAtStartOfLine()` check here?

Comment (hamishknight) [reply lib/Parse/ParseDecl.cpp]: @rintaro Sure thing! I've gone ahead and made that change.

Thanks for the review!

**Comment by hamishknight** [comment]
https://github.com/apple/llvm-project/pull/712

@swift-ci please test

**Comment by hamishknight** [comment]
https://github.com/apple/llvm-project/pull/712

@swift-ci please test

File: lib/Parse/ParseDecl.cpp (hunk 1 above)
Comment (hamishknight) [reply lib/Parse/ParseDecl.cpp]: I've gone ahead and elaborated a little on the comment

**Comment by hamishknight** [comment]
Rebased to resolve a merge conflict

**Comment by hamishknight** [comment]
https://github.com/apple/llvm-project/pull/712

@swift-ci please test

**Comment by hamishknight** [comment]
https://github.com/apple/llvm-project/pull/712

@swift-ci please test

**Comment by hamishknight** [comment]
@slavapestov If you're happy with this, would you mind doing the honors and merging it alongside the LLVM patches?

https://github.com/apple/llvm-project/pull/712
https://github.com/apple/llvm-project/pull/716
https://github.com/apple/llvm-project/pull/717

(I still need to get commit access to llvm-project)

**Comment by hamishknight** [comment]
https://github.com/apple/llvm-project/pull/712

@swift-ci please test

**Pull request merged**